codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestCompressor.java

```diff
@@ -0,0 +1,361 @@
+ *     }
+ *   });
+ * </pre>
+ */
+public class HttpRequestCompressor extends ChannelOutboundHandlerAdapter {
+
+    /**
+     * default encoding. used if no preferred encoding is set or it is not available.
+     */
+    public static final String DEFAULT_ENCODING = "gzip";
+    private static final List<String> SUPPORTED_ENCODINGS = Arrays.asList(new String[] {
+        "br",
+        "zstd",
+        "snappy",
+        "deflate",
+        "gzip"
+    });
+    private static final InternalLogger log = InternalLoggerFactory.getInstance(HttpRequestCompressor.class);
+    private final int contentSizeThreshold;
+    private final String encoding;
+    private final Supplier<MessageToByteEncoder<ByteBuf>> encoderFactory;
+
+    /**
+     * Key of the attribute that will hold the original request that is needed later
+     * in order to create a {@link FullHttpRequest}
+     */
+    private final AttributeKey<HttpRequest> reqAttrKey = AttributeKey.valueOf(HttpRequestCompressor.class, "req");
+
+    /**
+     * Key of the buffer that will hold all the payload chunks that we have to compress later
+     */
+    private final AttributeKey<ByteBuf> bufAttrKey = AttributeKey.valueOf(HttpRequestCompressor.class, "buf");
+
+    /**
+     * new instance using the defaults.
+     * shortcut for {@code new HttpRequestCompressor(DEFAULT_ENCODING)}
+     * @see #DEFAULT_ENCODING
+     * @see #HttpRequestCompressor(java.lang.String)
+     */
+    public HttpRequestCompressor() {
+        this(DEFAULT_ENCODING);
+    }
+
+    /**
+     * new instance using the preferred encoding.
+     * shortcut for {@code new HttpRequestCompressor(preferredEncoding, 0)}
+     * @param preferredEncoding the preferred encoding.
+     * if unavailable, the default encoding {@link #DEFAULT_ENCODING} will be used.
+     * @see #HttpRequestCompressor(java.lang.String, int)
+     */
+    public HttpRequestCompressor(String preferredEncoding) {
+        this(preferredEncoding, 0);
+    }
+
+    /**
+     * new instance using the preferred encoding and the given threshold.
+     * shortcut for {@code new HttpRequestCompressor(preferredEncoding, contentSizeThreshold, null)}
+     * @param preferredEncoding the preferred encoding.
+     * if unavailable, the default encoding {@link #DEFAULT_ENCODING} will be used.
+     * @param contentSizeThreshold the size in byte the http body must have before compressing the request
+     * @see #HttpRequestCompressor(java.lang.String, int, io.netty.handler.codec.compression.CompressionOptions...)
+     */
+    public HttpRequestCompressor(String preferredEncoding, int contentSizeThreshold) {
+        this(preferredEncoding, contentSizeThreshold, (CompressionOptions[]) null);
+    }
+
+    /**
+     * new instance using the preferred encoding, threshold and compression options.
+     * @param preferredEncoding the preferred encoding.
+     * if unavailable, the default encoding {@link #DEFAULT_ENCODING} will be used.
```

Comment:
If the preferred encoding is unsupported or unavailable I think we should throw instead. On the Netty level I think it's preferable to be explicit about things.